### PR TITLE
Add granular DiffTypes to the diffProperty decorator

### PR DESCRIPTION
**Pull request by matt-gadd**
**Type:** bug / feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

**Description:**
This resolves partially https://github.com/dojo/widget-core/issues/404. It doesn't change any of the fallback diff behaviour, but now adds diffTypes that can be applied per property at a class level, as well as supporting the existing `diffProperty` behaviour. The new usage of `diffProperty` at a class level looks like:

```ts
@diffProperty('foo', DiffType.SHALLOW)
@diffProperty('bar', DiffType.REFERENCE)
class MyWidget extends WidgetBase<any> {
}
```

Note that the decorators are stackable and take a DiffType of the following:

```
CUSTOM,
ALWAYS,
IGNORE,
REFERENCE,
SHALLOW,
AUTO
```

The `AUTO` type will select a diff behaviour based on the following:
* If its a function, ignore diffing completely
* If its an object or array, diff the values shallowly by reference
* otherwise diff by reference

### CODE REVIEW

File: src/diff.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+			}
+			else if (isObjectOrArray(newProperty)) {
+				result = shallow(previousProperty, newProperty);
+			}
+			else {
+				result = reference(previousProperty, newProperty);
+			}
+		break;
+		default:
+			result = {
```
Comment (agubler) [inline src/diff.ts]: Perhaps we should emit or log a warning if a user accidentally gets here.

File: src/WidgetBase.ts [hunk 2]
```diff
@@ -55,9 +59,14 @@ export function afterRender(target: any, propertyKey: string, descriptor: Proper
...
  */
-export function diffProperty(propertyName: string) {
-	return function (target: any, propertyKey: string, descriptor: PropertyDescriptor) {
-		target.addDecorator('diffProperty', { propertyName, diffFunction: target[propertyKey] });
+export function diffProperty(propertyName: string, diffType = DiffType.CUSTOM, diffFunction?: Function) {
+	return function (target: any, propertyKey?: string, descriptor?: PropertyDescriptor) {
+		if (diffType === DiffType.CUSTOM && propertyKey && descriptor) {
+			target.addDecorator('diffProperty', { propertyName, diffType, diffFunction: target[propertyKey] });
+		}
+		else if (diffType && !propertyKey && !descriptor) {
```
Comment (agubler) [inline src/WidgetBase.ts]: We don't cover the `else` branch (i'm not sure it is ever a realistic outcome)

File: src/diff.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+function ignore(previousProperty: any, newProperty: any): PropertyChangeRecord {
+	return {
+		changed: false,
+		value: newProperty
+	};
+}
+
+function custom(previousProperty: any, newProperty: any, meta: any): PropertyChangeRecord {
```
Comment (matt-gadd) [inline src/diff.ts]: the `meta` is meant to be a generic value bag. we only have one usage of it at the moment which does have `diffFunction` and `scope`, but thats extremely coupled to the existing functionality of `diffProperty`. I think we'll revisit the decorator functions in https://github.com/dojo/widget-core/issues/402, so I think we should leave as is for now (given its internal anyway).

File: src/WidgetBase.ts [hunk 4]
```diff
@@ -34,7 +37,8 @@ interface WidgetCacheWrapper {
  */
 interface DiffPropertyConfig {
 	propertyName: string;
-	diffFunction: Function;
+	diffType: DiffType;
+	diffFunction?: Function;
```
Comment (agubler) [inline src/WidgetBase.ts]: Sorry I meant `diffFunction` not `diffProperty`.

File: src/diff.ts (hunk 3 above)
Comment (agubler) [inline src/diff.ts]: Understood

**Pull request merged**